src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,192 @@
+            System.arraycopy(elements, 0, newArray, 0, elements.length);
+            elements = newArray;
+        }
+    }
+
     @Override
-    public void add(T value) {
+    public boolean add(T element) {
+        ensureCapacity();
+        elements[size++] = element;
+        return true;
+    }
+
+    @Override
+    public void add(int index, T element) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Індекс " + index + " виходить за межі списку.");
+        }
+        ensureCapacity();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = element;
+        size++;
+    }
 
+    @Override
+    public T get(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " to go beyond the list");
+        }
+        return (T) elements[index];
     }
 
     @Override
-    public void add(T value, int index) {
+    public T remove(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " to go beyond the list");
+        }
+        T removedElement = (T) elements[index];
+        System.arraycopy(elements, index + 1,
+                elements, index, size - index - 1);
+        elements[--size] = null;
+        return removedElement;
+    }
+
+    public boolean remove(Object element) {
+        for (int i = 0; i < size; i++) {
+            // Якщо обидва значення null або якщо елементи рівні
+            if (element == null ? elements[i] == null : element.equals(elements[i])) {
+                remove(i);
+                return true;
+            }
+        }
+        throw new NoSuchElementException(
+                "Element not found.");
+    }
 
+    @Override
+    public int size() {
+        return size;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public boolean isEmpty() {
+        return size == 0;
+    }
 
+    public boolean contains(Object o) {
+        for (int i = 0; i < size; i++) {
+            if (o == null ? elements[i] == null : o.equals(elements[i])) {
+                return true;
+            }
+        }
+        return false;
     }
 
     @Override
-    public T get(int index) {
+    public Iterator<T> iterator() {
         return null;
     }
 
+    public void clear() {
+        for (int i = 0; i < size; i++) {
+            elements[i] = null;
+        }
+        size = 0;
+    }
+
     @Override
-    public void set(T value, int index) {
+    public T set(int index, T element) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index + " to go beyond the list");
+        }
+        T oldValue = (T) elements[index];
+        elements[index] = element;
+        return oldValue;
```

Comment:
The index check logic is repeated in multiple methods (add, get, remove). Consider moving it to a separate private method to avoid code duplication and adhere to the DRY principle.